project/SUNGOD3/README.rst

```diff
@@ -0,0 +1,154 @@
+
+1. Improving search speed beyond basic linear search, especially for 
+   high-dimensional vector spaces common in RAG systems.
+2. Developing new similarity metrics that may be more suitable for 
+   specific RAG applications.
+3. Creating a flexible, extensible framework that allows for easy 
+   experimentation with different search algorithms and metrics.
+4. Balancing speed, accuracy, and memory usage in vector search operations.
+
+Prospective Users
+-----------------
+This project will benefit:
+
+1. Researchers and developers working on RAG systems.
+2. Data scientists and machine learning engineers dealing with large-scale 
+   similarity search problems.
+3. Information retrieval specialists looking for customizable vector 
+   search solutions.
+4. Anyone working with high-dimensional data requiring efficient similarity 
+   search capabilities.
+5. Graduate students who lack a graduation thesis.
+
+System Architecture
+-------------------
+The AdvancedVectorSearch system will consist of the following components:
+
+1. Dataset Module: For generating and managing large-scale, high-dimensional 
+   vector datasets for testing and benchmarking.
+2. Index Module: Implementing various similarity comparison functions and 
+   indexing structures.
+3. Search Module: Providing add and search functionalities with support for 
+   different search algorithms.
+4. Benchmark Module: For comparing the implemented methods against linear 
+   search and potentially FAISS in terms of speed, accuracy, and memory usage.
+
+Index Module
+------------
+
+The Index Module is responsible for implementing various similarity comparison 
+functions and indexing structures. It supports basic operations for adding, 
+searching, and removing vectors from the index. The basic index structure 
+includes the following:
+
+- **vector<float>** : Stores the vector data.
+- **int** : Stores the vector dimension.
+- **int** : Stores the number of vectors in the index.
+- **string** : Unique identifier for each vector.
+
+The module provides the following methods:
+
+- **add(vector)** : Adds vectors to the index.
+- **search(query_vector)** : Searches the index given a query vector.
+- **remove(vector)** : Removes vectors from the index.
+
+Supported Index Types
+---------------------
+
+The Index Module supports the following types of indices:
+
+1. **Flat Index**: 
+   A simple index that stores vectors in memory and performs an 
+   exhaustive search.
+
+2. **IVF (Inverted File) Index**: 
+   An index that partitions the vector space into clusters for 
+   faster search.
+
+   This method provides a more efficient search process compared 
+   to flat indices, especially for large-scale datasets.
+
+The system will be implemented in C++ for core functionalities, with Python 
+bindings for ease of use.
+
+API Description
+---------------
+
+The AdvancedVectorSearch API will include:
+
+1. Dataset Management
+2. Index Operations
+3. Search Operations
+4. Benchmark
```

Comment:
I think in this section, you should talk about how you design the architecture in your system.

For example, you need dataset management in your application. Do you want to do a singleton class with some container like std::vector on it? What part do you want to work with C++ and optimize the speed of it?

It will be great if your clarify how you develop in this part.

Reply:
I'm not sure which implementation method is more suitable for this part. I hope I can find time to try it out before deciding.

Reply:
It may be danger if you just implement it without architecture design since the implementation may not fully consider your function requirement. We may discuss your idea and find a proper architecture design to implement your application.

Reply:
You will find it pretty challenging to time accurately on github actions.  It's a good experience to try to do it if you have not.